### PR TITLE
Ofer dev

### PR DESCRIPTION
**Pull request by helmanofer**
it is a work in progress
The pull request contains QuadgramCollocationFinder and QuadgramAssocMeasures

### CODE REVIEW

File: nltk/collocations.py [hunk 1]
```diff
@@ -64,19 +64,17 @@ def from_documents(cls, documents):
 
     @staticmethod
     def _ngram_freqdist(words, n):
-        return FreqDist(tuple(words[i:i+n]) for i in range(len(words)-1))
+        return FreqDist(tuple(words[i:i + n]) for i in range(len(words) - 1))
```
Comment (jnothman) [inline nltk/collocations.py]: FWIW, I find `i:i+n` clearer than `i:i + n` (and PEP8 is equivocal)

File: nltk/collocations.py [hunk 2]
```diff
@@ -64,19 +64,17 @@ def from_documents(cls, documents):
...
         """
-        items = list(self.ngram_fd.items())
-        for ngram, freq in items:
-            if fn(ngram, freq):
-                try:
-                    del self.ngram_fd[ngram]
-                except KeyError:
-                    pass
+        tmp_ngram = FreqDist()
+        for ngram, freq in self.ngram_fd.items():
```
Comment (jnothman) [inline nltk/collocations.py]: you can use `nltk.compat.iteritems` here.

File: nltk/collocations.py [hunk 3]
```diff
@@ -235,6 +237,84 @@ def score_ngram(self, score_fn, w1, w2, w3):
                         n_all)
 
 
+class QuadgramCollocationFinder(AbstractCollocationFinder):
+    """A tool for the finding and ranking of quadgram collocations or other association measures.
```
Comment (jnothman) [inline nltk/collocations.py:242]: please update this docstring

File: nltk/collocations.py [hunk 4]
```diff
@@ -235,6 +237,84 @@ def score_ngram(self, score_fn, w1, w2, w3):
                         n_all)
 
 
+class QuadgramCollocationFinder(AbstractCollocationFinder):
```
Comment (jnothman) [inline nltk/collocations.py:241]: I wonder if it's better to make a generic >3gram collocation finder, for users to experiment with. (I expect that as one increases n, collocation is much more correlated with raw frequency.)

The counters correspond to `[''.join(tup) for tup in itertools.product(['x', 'i'], repeat=4) if tup[0] == 'i']`, in which `word_fd` is the first and `ngram_fd` the last. You could calculate this list of masks, and build a `FreqDist` for each. Many association measures only use `word_fd` and `ngram_fd`, so depending on a user option, you could just ignore the other masks when collecting frequencies.

Comment (jnothman) [inline nltk/collocations.py:241]: Although determining a generic implementation of `score_ngram` and `_contingency` will take a bit of time.

**Comment by stevenbird** [comment]
Thanks @helmanofer and @jnothman. I'm happy to merge this. It might be nice to mention the existence of the QuadgramCollocationFinder in nltk/test/collocations.doctest.

**Comment by helmanofer** [comment]
Happy to contribute

**Comment by jnothman** [comment]
Thanks Steve. I'd not completely reviewed it yet (deadlines, deliberations, etc.), so hopefully it's correct...

**Pull request merged**